src/DMLQueryBuilder.php

```diff
@@ -135,74 +121,50 @@ public function upsert(
 
         foreach ($constraints as $constraint) {
             $constraintCondition = ['and'];
+            $columnNames = (array) $constraint->getColumnNames();
 
-            $columnNames = $constraint->getColumnNames() ?? [];
-
-            if (is_array($columnNames)) {
-                /** @psalm-var string[] $columnNames */
-                foreach ($columnNames as $name) {
-                    $quotedName = $this->quoter->quoteColumnName($name);
-                    $constraintCondition[] = "$quotedTableName.$quotedName=[EXCLUDED].$quotedName";
-                }
+            /** @psalm-var string[] $columnNames */
+            foreach ($columnNames as $name) {
+                $quotedName = $this->quoter->quoteColumnName($name);
+                $constraintCondition[] = "$quotedTableName.$quotedName=[EXCLUDED].$quotedName";
             }
 
             $onCondition[] = $constraintCondition;
         }
 
         $on = $this->queryBuilder->buildCondition($onCondition, $params);
 
-        /** @psalm-var string[] $placeholders */
         [, $placeholders, $values, $params] = $this->prepareInsertValues($table, $insertColumns, $params);
-        $mergeSql = 'MERGE ' . $this->quoter->quoteTableName($table) . ' WITH (HOLDLOCK) '
-            . 'USING (' . (!empty($placeholders)
-            ? 'VALUES (' . implode(', ', $placeholders) . ')'
-            : ltrim((string) $values, ' ')) . ') AS [EXCLUDED] (' . implode(', ', $insertNames) . ') ' . "ON ($on)";
-        $insertValues = [];
 
-        foreach ($insertNames as $name) {
-            $quotedName = $this->quoter->quoteColumnName($name);
+        $mergeSql = 'MERGE ' . $quotedTableName . ' WITH (HOLDLOCK) USING ('
+            . (!empty($placeholders) ? 'VALUES (' . implode(', ', $placeholders) . ')' : $values)
+            . ') AS [EXCLUDED] (' . implode(', ', $insertNames) . ') ' . "ON ($on)";
 
-            if (strrpos($quotedName, '.') === false) {
-                $quotedName = '[EXCLUDED].' . $quotedName;
-            }
+        $insertValues = [];
 
-            $insertValues[] = $quotedName;
+        foreach ($insertNames as $quotedName) {
+            $insertValues[] = '[EXCLUDED].' . $quotedName;
         }
 
-        $insertSql = 'INSERT (' . implode(', ', $insertNames) . ')' . ' VALUES (' . implode(', ', $insertValues) . ')';
+        $insertSql = 'INSERT (' . implode(', ', $insertNames) . ') VALUES (' . implode(', ', $insertValues) . ')';
 
-        if ($updateNames === []) {
+        if ($updateColumns === false || $updateNames === []) {
```

Comment:
```suggestion
        if (empty($updateColumns)) {
```

It's a little more productive.

Reply:
not correct suggestion